pkg/loadtest/loadtest.go

```diff
@@ -0,0 +1,140 @@
+package loadtest
+
+import (
+	"context"
+	cryptoRand "crypto/rand"
+	"sync"
+	"time"
+
+	"github.com/xmtp/xmtpd/pkg/blockchain"
+	"github.com/xmtp/xmtpd/pkg/config"
+	"github.com/xmtp/xmtpd/pkg/db"
+	"go.uber.org/zap"
+)
+
+func LoadTest(
+	logger *zap.Logger,
+	options *config.LoadTestOptions,
+	contractsOptions *config.ContractsOptions,
+) error {
+	ctx := context.Background()
+	chainClient, err := blockchain.NewClient(ctx, contractsOptions.RpcUrl)
+	if err != nil {
+		return err
+	}
+
+	signer, err := blockchain.NewPrivateKeySigner(
+		options.PrivateKey,
+		contractsOptions.ChainID,
+	)
+
+	if err != nil {
+		return err
+	}
+
+	dbConn, err := db.NewNamespacedDB(
+		ctx,
+		logger,
+		options.DbConnectionString,
+		"loadtest",
+		5*time.Second,
+		1*time.Second,
+	)
+	if err != nil {
+		return err
+	}
+
+	nonceManager := blockchain.NewSQLBackedNonceManager(dbConn, logger)
+	publisher, err := blockchain.NewBlockchainPublisher(
+		ctx,
+		logger,
+		chainClient,
+		signer,
+		*contractsOptions,
+		nonceManager,
+	)
+	if err != nil {
+		logger.Fatal("could not create blockchain publisher", zap.Error(err))
+	}
+
+	doLoadTest(ctx, logger, publisher, options)
+
+	return nil
+}
+
+func doLoadTest(
+	ctx context.Context,
+	logger *zap.Logger,
+	publisher *blockchain.BlockchainPublisher,
+	options *config.LoadTestOptions,
+) {
+	groupID := randomGroupID()
+
+	// Calculate total messages to send
+	totalMessages := options.MessagesPerSecond * options.Duration
+
+	// Create a ticker to control the rate of messages
+	interval := time.Second / time.Duration(options.MessagesPerSecond)
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	logger.Info("Starting load test",
+		zap.Int("messagesPerSecond", options.MessagesPerSecond),
+		zap.Int("duration", options.Duration),
+		zap.Int("totalMessages", totalMessages),
+		zap.Int("messageSize", options.MessageSizeBytes))
+
+	// Track metrics
+	var sentCount int
+	startTime := time.Now()
+
+	var wg sync.WaitGroup
+
+	// Send messages at the specified rate
+	for i := range totalMessages {
+		<-ticker.C
+		wg.Add(1)
+		// Send a message
+		go func(messageNumber int) {
+			defer wg.Done()
+			if _, err := publisher.PublishGroupMessage(ctx, groupID, randomBytes(options.MessageSizeBytes)); err != nil {
+				logger.Error(
+					"Failed to publish message",
+					zap.Error(err),
+					zap.Int("messageNumber", messageNumber+1),
+				)
+			}
+		}(i)
```

Comment:
can we return an error code all the way to the client? That way you can do something along the lines of `while (loadtest); do...`

Reply:
I don't want to give up on a single error. But I could turn them into an array and return the whole thing when the test is done.

Reply:
either that or just return "has failed yes/no". I agree that early exit is undesired.